model/jpa/src/main/java/org/jboss/aerogear/unifiedpush/jpa/dao/impl/JPAPushMessageInformationDao.java

```diff
@@ -41,8 +41,27 @@
 
     @Override
     public List<PushMessageInformation> findAllForPushApplication(String pushApplicationId, boolean ascending) {
-        return createQuery("select pmi from PushMessageInformation pmi where pmi.pushApplicationId = :pushApplicationId ORDER BY pmi.submitDate " + ascendingOrDescending(ascending))
-                .setParameter("pushApplicationId", pushApplicationId).getResultList();
+        return findAllForPushApplicationByParams(pushApplicationId, null, ascending, null, null);
+    }
+
+    @Override
+    public List<PushMessageInformation> findAllForPushApplicationByParams(String pushApplicationId, String search, boolean ascending, Integer page, Integer pageSize) {
+        String baseQuery = "from PushMessageInformation pmi where pmi.pushApplicationId = :pushApplicationId";
+        if (search != null) {
+            baseQuery += " AND pmi.rawJsonMessage LIKE :search";
+        }
+        final String queryJPQL = "select pmi " + baseQuery + " ORDER BY pmi.submitDate " + ascendingOrDescending(ascending);
+
+        TypedQuery<PushMessageInformation> typedQuery = createQuery(queryJPQL)
+                .setParameter("pushApplicationId", pushApplicationId);
+        if (search != null) {
+            typedQuery.setParameter("search", "%" + search + "%");
```

Comment:
This one may pose an issue, as the wildcard operator at the beginning of the search term likely will trigger a full-table scan. Can you resort to (vendor-specific) full-text search functions? Or you use Hibernate Search and Lucene :)

But it's all a bit guess work on my side, would be interesting to see some profiling results of where your execution actually is spent and how execution plans of the queries look like.